[Gstreamer][Westeros] Bring back westeros autoplug code for playbin3

calvaris opened the pull request:
#### 2ffd552e8048df88b9366b4a4c3c7d45e9850ea2
<pre>
[Gstreamer][Westeros] Bring back westeros autoplug code for playbin3
<a href="https://bugs.webkit.org/show_bug.cgi?id=263714">https://bugs.webkit.org/show_bug.cgi?id=263714</a>

Reviewed by Philippe Normand.

To autoplug westeros sink in playbin3 we need to specify the stop caps to ensure encoded data goes directly into the
sink.

This is a revert of # <a href="https://commits.webkit.org/266178@main.">https://commits.webkit.org/266178@main.</a>

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::syncOnClock):
(WebCore::MediaPlayerPrivateGStreamer::configureElementPlatformQuirks):

Canonical link: <a href="https://commits.webkit.org/269808@main">https://commits.webkit.org/269808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bfdaf1102c215c6275f6f4ab51d570a0d785d7c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24763 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25801 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21825 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23910 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24178 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22387 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23885 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20456 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26397 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21362 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27635 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21563 "Found 1 new API test failure: TestWebKitAPI.ServiceWorker.WindowClientNavigate (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25399 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1070 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18773 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1049 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5649 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1483 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1360 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2306,12 +2304,40 @@ void MediaPlayerPrivateGStreamer::configureElementPlatformQuirks(GstElement* ele
+    if (!m_isLegacyPlaybin && westerosSinkCaps && g_str_has_prefix(GST_ELEMENT_NAME(element), "uridecodebin3")) {
+        GRefPtr<GstCaps> defaultCaps;
+        g_object_get(element, "caps", &defaultCaps.outPtr(), NULL);
+        defaultCaps = adoptGRef(gst_caps_merge(gst_caps_ref(westerosSinkCaps), defaultCaps.leakRef()));
```
Isn't westerosSinkCaps leaked here? Why ref it?

calvaris replied:
`gst_caps_merge` takes two _transfer full_ and returns _transfer full_. There you can "leak" or not because it is going to be handled gracefully by the smart pointer, but "leak" is ok unless I am missing anything.

philn replied:
I'm not talking about `defaultCaps.leakRef()`, which is OK.

I'm wondering about `gst_caps_ref(westerosSinkCaps)`... its ref count becomes 2, gst_caps_merge decreases it to 1, and after the gst_caps_merge() call there is no unref of `westerosSinkCaps`, unless I am missing anything.

calvaris replied:
`westerosSinkCaps` is static.

calvaris replied:
so we want to keep a reference alive for the next run

philn replied:
Ahh static, that's i missed. thx

calvaris committed:
[Gstreamer][Westeros] Bring back westeros autoplug code for playbin3
https://bugs.webkit.org/show_bug.cgi?id=263714

Reviewed by Philippe Normand.

To autoplug westeros sink in playbin3 we need to specify the stop caps to ensure encoded data goes directly into the
sink.

This is a revert of # https://commits.webkit.org/266178@main.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::syncOnClock):
(WebCore::MediaPlayerPrivateGStreamer::configureElementPlatformQuirks):

Canonical link: https://commits.webkit.org/269808@main
